### PR TITLE
 terraform and vpc-app-deploy

### PR DESCRIPTION
**Pull request by powellquiring**
terraform for bastion and pub-priv along with the app.

### CODE REVIEW

File: vpc-secure-management-bastion-server/tf/variables.tf [hunk 1]
```diff
@@ -0,0 +1,19 @@
+# see export.template for these:
+variable bluemix_api_key { }
```
Comment (l2fprod) [inline vpc-secure-management-bastion-server/tf/variables.tf]: use ibmcloud_api_key

File: export.template [hunk 2]
```diff
@@ -0,0 +1,14 @@
+# These environment variables are in terraform format
+# They will be converted appropriately to the other formats required
+
+# Download an API key and put it in this variable, it will look something like this:
+# TF_VAR_bluemix_api_key=r3aaaaI_nLeAVraafjKyIs5xkQnFWK5-doHgommVmooA
+export TF_VAR_bluemix_api_key=
```
[the remark this answers is not in the document]
Comment (powellquiring) [reply export.template]: bug

File: vpc-app-deploy/README.md [hunk 3]
```diff
@@ -0,0 +1,29 @@
+# Application deploy onto a vpc instance
+How to install software and files onto a Virtual Private Cloud, VPC, virtual server instance, VSI.
+
+Here is the full [solution tutorial](https://cloud.ibm.com/docs/vpc-on-classic?topic=solution-tutorials-vpc-app-deploy)
```
[the remark this answers is not in the document]
Comment (powellquiring) [reply vpc-app-deploy/README.md]: bug

File: export.template [hunk 4]
```diff
@@ -0,0 +1,14 @@
+# These environment variables are in terraform format
```
[the remark this answers is not in the document]
Comment (powellquiring) [reply export.template]: still trying to determine where and how many times this should be in the repo.  Added a check box:
https://github.ibm.com/Bluemix/cloud-portfolio-solutions/issues/1350#issuecomment-12038947

File: vpc-app-deploy/shared/cloud-config.yaml [hunk 5]
```diff
@@ -0,0 +1,51 @@
+#cloud-config
+bootcmd:
+- sed -i -e "s/mirrors.service.networklayer.com\|mirrors.service.softlayer.com/mirrors.adn.networklayer.com/g" /etc/yum.repos.d/*.repo*
```
[the remark this answers is not in the document]
Comment (powellquiring) [reply vpc-app-deploy/shared/cloud-config.yaml]: https://github.ibm.com/riaas/defects/issues/1834

File: vpc-app-deploy/Makefile [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+	cd $(TF); terraform init
+	cd $(TF); terraform apply -auto-approve
+test_public:
+	bash -x ./test_provision.bash $$(cd tf; terraform output FRONT_IP_ADDRESS) $(FRONT_INDEX) $(BOOTSTRAPMIN)
+test_private:
+	bash -x ./test_provision.bash $$(cd tf; terraform output BACK_NIC_IP) $(BACK_INDEX) $(BOOTSTRAPMIN) "root@$$(cd tf; terraform output FRONT_NIC_IP) -o ProxyJump=root@$$(cd tf; terraform output BASTION_IP_ADDRESS)"
+destroy:
+	cd $(TF); terraform destroy
+
+#all_cli_provision:
```
[the remark this answers is not in the document]
Comment (powellquiring) [reply vpc-app-deploy/Makefile]: dead code, cleaned up

File: vpc-secure-management-bastion-server/tf/export.template [hunk 7]
```diff
@@ -0,0 +1,11 @@
+#export TF_VAR_bluemix_api_key=r3aaaaI_nLeAVraafjKyIs5xkQnFWK5-doHgommVmooA
```
[the remark this answers is not in the document]
Comment (powellquiring) [reply vpc-secure-management-bastion-server/tf/export.template]: bug

File: vpc-secure-management-bastion-server/tf/variables.tf (hunk 1 above)
Comment (powellquiring) [reply vpc-secure-management-bastion-server/tf/variables.tf]: bug

File: vpc-public-app-private-backend/tf/export [hunk 8]
```diff
@@ -0,0 +1,11 @@
+#export TF_VAR_bluemix_api_key=r3aaaaI_nLeAVraafjKyIs5xkQnFWK5-doHgommVmooA
```
[the remark this answers is not in the document]
Comment (powellquiring) [reply vpc-public-app-private-backend/tf/export]: fixed

File: export.template (hunk 2 above)
[the remark this answers is not in the document]
Comment (powellquiring) [reply export.template]: fixed

File: vpc-app-deploy/export.template [hunk 9]
```diff
@@ -0,0 +1,11 @@
+#export TF_VAR_bluemix_api_key=r3aaaaI_nLeAVraafjKyIs5xkQnFWK5-doHgommVmooA
+# see https://cloud.ibm.com/docs/terraform?topic=terraform-getting-started to follow the instructions to get 
+# terraform installed and how to get the api key
+export TF_VAR_bluemix_api_key=<api key>
```
[the remark this answers is not in the document]
Comment (powellquiring) [reply vpc-app-deploy/export.template]: bug/fixed

File: vpc-public-app-private-backend/vpc-pubpriv-create-with-bastion.sh [hunk 10]
```diff
@@ -33,15 +36,27 @@ else
...
+fi
+
+
+if [ ! -z "$7" ]; then 
+    resource_file=$7
+fi
 
 export basename="vpc-pubpriv"
-export UbuntuImage=$(ibmcloud is images --json | jq -r '.[] | select (.name=="ubuntu-18.04-amd64") | .id')
+export UbuntuImage=$(ibmcloud is images --json | jq -r '.[] | select (.name=="centos-7.x-amd64") | .id')
```
[the remark this answers is not in the document]
Comment (l2fprod) [reply vpc-public-app-private-backend/vpc-pubpriv-create-with-bastion.sh]: fixed

File: vpc-app-deploy/test_provision.bash [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+      else
+        hi=$(ssh -F shared/ssh.config $bastion_ip curl -s $host_ip/$bootstrapfile)
+      fi
+      if [ "x$hi" = "x$expectingBootstrap" ]; then
+        echo success: provision of file from on premises worked and was replaced with the following contents:
+        echo $hi
+        exit 0
+      else
+        echo $hi
+        echo "fail: terraform provision does not work, expecting $expectingBootstrap but got the stuff above intead"
```
[the remark this answers is not in the document]
Comment (l2fprod) [reply vpc-app-deploy/test_provision.bash]: fixed

File: vpc-public-app-private-backend/README.md [hunk 12]
```diff
@@ -37,6 +29,38 @@ The following **named** resources are created by the script above:
...
+```
+
+To create the resources:
+
+```
+cp export.template export
+vi export; # make the changes suggested
+source export
+cd tf
+terraform apply
```
[the remark this answers is not in the document]
Comment (l2fprod) [reply vpc-public-app-private-backend/README.md]: fixed

**Pull request merged**